apiserver/facades/client/application/application_unit_test.go

```diff
@@ -420,37 +240,57 @@ func (s *ApplicationSuite) TestUpdateCAASApplicationSettings(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 	appCfgSchema, defaults, err = application.AddTrustSchemaAndDefaults(appCfgSchema, defaults)
 	c.Assert(err, jc.ErrorIsNil)
-
 	appCfg, err := coreconfig.NewConfig(map[string]interface{}{
 		"juju-external-hostname": "foo",
 	}, appCfgSchema, nil)
 	c.Assert(err, jc.ErrorIsNil)
+	app.EXPECT().UpdateApplicationConfig(appCfg.Attributes(), []string(nil), appCfgSchema, defaults).Return(nil)
 
-	pgApp.CheckCall(c, 3, "UpdateApplicationConfig", appCfg.Attributes(), []string(nil), appCfgSchema, defaults)
-}
+	s.backend.EXPECT().Application("postgresql").Return(app, nil)
 
-func (s *ApplicationSuite) TestSetCAASConfigSettings(c *gc.C) {
-	s.model.modelType = state.ModelTypeCAAS
-	s.setAPIUser(c, names.NewUserTag("admin"))
-	s.backend.charm = &mockCharm{
-		meta: &charm.Meta{
-			Deployment: &charm.Deployment{
-				DeploymentMode: charm.ModeOperator,
-			},
-		},
+	// Update settings for the application.
+	args := params.ApplicationUpdate{
+		ApplicationName: "postgresql",
+		SettingsYAML:    "postgresql:\n  stringOption: bar\n  juju-external-hostname: foo",
 	}
+	api := &application.APIv12{&application.APIv13{s.api}}
+	err = api.Update(args)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *ApplicationSuite) TestUpdateCAASApplicationSettingsInIAASModelTriggersError(c *gc.C) {
+	ctrl := s.setup(c, state.ModelTypeIAAS)
+	defer ctrl.Finish()
+
+	ch := s.expectCharm(ctrl, nil, nil, &charm.Config{Options: map[string]charm.Option{
+		"stringOption": {Type: "string"},
+		"intOption":    {Type: "int", Default: int(123)},
+	}})
+
+	app := s.expectApplicationWithCharm(ctrl, ch, "postgresql", "quantal")
+	s.backend.EXPECT().Application("postgresql").Return(app, nil)
 
 	// Update settings for the application.
-	args := params.ConfigSetArgs{Args: []params.ConfigSet{{
+	args := params.ApplicationUpdate{
 		ApplicationName: "postgresql",
-		ConfigYAML:      "postgresql:\n  stringOption: bar\n  juju-external-hostname: foo",
-	}}}
-	results, err := s.api.SetConfigs(args)
-	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(results.Results, gc.DeepEquals, []params.ErrorResult{{}})
+		SettingsYAML:    "postgresql:\n  stringOption: bar\n  juju-external-hostname: foo",
+	}
+	api := &application.APIv12{&application.APIv13{s.api}}
+	err := api.Update(args)
+	c.Assert(err, gc.ErrorMatches, `.*unknown option "juju-external-hostname"`, gc.Commentf("expected to get an error when attempting to set CAAS-specific app setting in IAAS model"))
+}
+
+func (s *ApplicationSuite) TestSetCAASConfigSettings(c *gc.C) {
```

Comment:
Good shout. All cases of UpdateApplicationConfig are identical, with one exception. I have extracted that out into a method

Reply:
Looks like this wasn't _quite_ what you were referring two. I'm not too keen on extracting the whole setup as a function (which was common in the old tests), unless there's some meaningful reason why it should be, which I'm not convinced there is here (there often isn't in tests I find). But that still doesn't address the problem that there's a lot of repetition, so instead I've extracted out some of the common steps, namely:
- expectDefaultCharm
- expectDefaultApplication
- expectUpdateApplicationConfig

Reply:
If many lines of setup are copied from test to test, they are not essential to the content of what the test is testing. In the future it becomes harder to change and harder to understand what exactly is being tested.

Reply:
I disagree that copied lines necessarily mean they are non-essential and make things harder to change. Particularly in tests. 

I reckon there's a good chance that if a change need to be made within either these tests, it will only be required in one of these two tests, meaning to extract a function will make changes more difficult. Particularly since these tests test different function, one of which is a legacy function

But that said, as it stands there are only 4 lines of setup duplication (there was defo too much in the past so thank you for raising it), which read roughly like "you expect an app to exist called 'postgresql", which expects a call to `UpdateCharmConfig` and `UpdateApplicationConfig`", which imo communicates the essence of the test _fairly_ well, and is at a good level of abstraction

Happy to discuss this more with you if need be :grin: